app/routes/atlas/reduceDimensions.ts

```diff
@@ -0,0 +1,145 @@
+import { TSNE } from '@keckelt/tsne'
+import { UMAP } from 'umap-js'
+import { lapJv } from './lap'
+import { getSquaredEuclideanDistance } from './distance'
+
+export type Point = {
+  x: number
+  y: number
+}
+
+export function reduceDimensions(
+  data: number[][],
+  alg: 'tsne' | 'umap',
+): number[][] {
+  // https://youtu.be/NEaUSP4YerM
+  if (alg === 'tsne') {
+    let tsne = new TSNE({
+      epsilon: 10, // learning rate (10 = default)
+      perplexity: 30, // how many neighbors does each point influence  (30 = default)
+      dim: 2, // dimensionality of the embedding (2 = default)
+    })
+    tsne.initDataRaw(data)
+    let iterations = 100
+
+    for (let i = 0; i < iterations; i++) {
+      tsne.step()
+    }
+
+    return tsne.getSolution()
+  }
+
+  // https://youtu.be/eN0wFzBA4Sc
+  if (alg === 'umap') {
+    let umap = new UMAP()
+    return umap.fit(data)
+  }
+
+  return []
+}
+
+type GridSize = { width: number; height: number }
+
+export function toGridIndexesWithGaps(points: number[][], gridSize: GridSize) {
+  let normalizedPoints = normalizeCoordinates(points)
+
+  return normalizedPoints.map((point) => {
+    const gridX = Math.floor(point[0] * (gridSize.width - 1))
+    const gridY = Math.floor(point[1] * (gridSize.height - 1))
+
+    return [gridX, gridY]
+  })
+}
+
+export function toGridIndexesNoGaps(points: number[][], gridSize: GridSize) {
```

Comment:
Added a new function aiming to solve gaps in the grid. 
![image](https://github.com/vladinator1000/remix-drizzle-d1/assets/1226564/546b6142-56fd-434b-83c0-b565f8c5dd7c)


My current implementation still leaves gaps at the bottom though 😭 
![image](https://github.com/vladinator1000/remix-drizzle-d1/assets/1226564/4b8676e3-277e-40af-a688-935ff1a78571)